### PR TITLE
[ServiceBus] Track1 - Dead Letter Fix

### PR DESCRIPTION
**Pull request by yunhaoling**
**pending uamqp v1.2.8 release**

https://github.com/Azure/azure-uamqp-python/pull/160

### CODE REVIEW

File: sdk/servicebus/azure-servicebus/azure/servicebus/aio/async_message.py [hunk 1]
```diff
@@ -89,16 +94,34 @@ async def dead_letter(self, description=None):
...
         :raises: ~azure.servicebus.common.errors.MessageLockExpired if message lock has already expired.
         :raises: ~azure.servicebus.common.errors.SessionLockExpired if session lock has already expired.
         :raises: ~azure.servicebus.common.errors.MessageSettleFailed if message settle operation fails.
         """
         self._is_live('reject')
+
+        info = None
+        if description:
+            info = {
+                RECEIVER_LINK_DEAD_LETTER_REASON: reason if reason else description,
```
Comment (KieranBrantnerMagee) [inline sdk/servicebus/azure-servicebus/azure/servicebus/aio/async_message.py:110]:
```suggestion
                RECEIVER_LINK_DEAD_LETTER_REASON: reason or description,
```
Silly nit, wouldn't even suggest this if I couldn't make the change for you and have it be one click.  (Just some cute syntactic sugar for condensing that sort of ternary)

File: sdk/servicebus/azure-servicebus/azure/servicebus/aio/async_message.py [hunk 2]
```diff
@@ -193,15 +216,16 @@ async def dead_letter(self, description=None):
...
+        :param str reason: The reason for dead-lettering the message. If `reason` is not set while `description` is
+         set, then `reason` would be set the same as `description`.
         :raises: ~azure.servicebus.common.errors.MessageAlreadySettled if the message has been settled.
         :raises: ~azure.servicebus.common.errors.MessageLockExpired if message lock has already expired.
         :raises: ~azure.servicebus.common.errors.MessageSettleFailed if message settle operation fails.
         """
         self._is_live('dead-letter')
         details = {
-            'deadletter-reason': str(description) if description else "",
+            'deadletter-reason': str(reason) if reason else (str(description) if description else ""),
```
Comment (KieranBrantnerMagee) [inline sdk/servicebus/azure-servicebus/azure/servicebus/aio/async_message.py:228]: Is there a limitation where description as None would be problematic?  If so, should this be something we have above?  (as opposed to allowing the caller to pass None explicitly)

There's also the slight asymmetry where we have the details object populated with a deadletter-description even if reason is the only present parameter, whereas above we don't populate the description field.    If we're keeping this for consistency/backcompat I'm fine with that, consider this a note for normalizing these behaviors for T1.

File: sdk/servicebus/azure-servicebus/azure/servicebus/common/message.py [hunk 3]
```diff
@@ -337,16 +341,29 @@ def dead_letter(self, description=None):
...
         :raises: ~azure.servicebus.common.errors.MessageLockExpired if message lock has already expired.
         :raises: ~azure.servicebus.common.errors.SessionLockExpired if session lock has already expired.
         :raises: ~azure.servicebus.common.errors.MessageSettleFailed if message settle operation fails.
         """
         self._is_live('reject')
+
+        info = None
+        if description:
+            info = {
+                RECEIVER_LINK_DEAD_LETTER_REASON: reason if reason else description,
```
Comment (KieranBrantnerMagee) [inline sdk/servicebus/azure-servicebus/azure/servicebus/common/message.py:357]:
```suggestion
                RECEIVER_LINK_DEAD_LETTER_REASON: reason or description,
```

File: sdk/servicebus/azure-servicebus/azure/servicebus/aio/async_message.py (hunk 2 above)
Comment (yunhaoling) [reply sdk/servicebus/azure-servicebus/azure/servicebus/aio/async_message.py:228]: re1: no limitation on description being None. actually we even could accept this -- `details = { 'deadletter-reason': None, 'deadletter-reason': None} `.

re2: They're two separate fields which shouldn't interfere with each other. But here is mainly for consistency/backcompat support in T1.

**Pull request merged**